### PR TITLE
Fix #22140 - Add bech32 encoding/decoding

### LINKED ISSUE #22140: Add bech32 encoding/decoding
This encoding is used by nostr and some crypto currencies to encode keys and notes with a meaningful prefix and extend the hexadecimal wordability. This can be compared to base64 but it's just different :D 

https://github.com/ElementsProject/lightning/blob/master/common/bech32.c

### PR DESCRIPTION
**Pull request by W0nda**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

Add bech32 encoding/decoding but I don't know if I should integrate that to rax2 or not...

### CODE REVIEW

**Comment by trufae** [comment]
Maybe makes more sense to be in rahash2

**Comment by trufae** [comment]
Its probably better to have it as a crypto plugin for encoding. Also chk the linting issues to follow the indentation rules (space before tab) and having some tests would be great too.

good work!

**Comment by trufae** [comment]
Will be good to have the plugin , take a look at the crypto_base64.c . if you dont expose this api and use it directly in the plugin it wont need to be new-abi specific

**Comment by W0nda** [comment]
I'm implementing tests, next pushed commits will be:

1. tests
2. implementation of bech32 line base64 is

**Comment by W0nda** [comment]
The crypto plugins implementation isn't valid at all. Sorry for having pushing too soon.
I'll correct that and push tomorrow or next week.

**Comment by trufae** [comment]
sounds good. there's no hurry. ill cut the release before merging this

**Comment by trufae** [comment]
github decided to broke all ubuntu-latest images, i fixed that on master, can you rebase your patches on top of master?

**Comment by W0nda** [comment]
The encoding code as been tested and works but I don't know how to integrate it into r2. Each time I try to build it fails.

(The implem in crypto_bech32.c isn't totally finished but I tried to understand how to compile it first)

**Comment by trufae** [comment]
The header file is not there. Maybe you forgot to commit it?

File: libr/include/r_crypto.h
Comment (trufae) [inline libr/include/r_crypto.h:12]: this file doesnt exist

File: libr/crypto/p/crypto_bech32.h
Comment (trufae) [inline libr/crypto/p/crypto_bech32.h:40]: we dont want to expose new APIs that will break the abi contract. Can you make this static for now? actually its only purpose is to be used by the plugin so it shouldnt be an issue

File: test/unit/test_bech32.c
Comment (trufae) [inline test/unit/test_bech32.c:148]: static

File: test/unit/test_bech32.c
Comment (trufae) [inline test/unit/test_bech32.c:155]: these tests use an api that shouldnt be made public to protect the abi contract. so you can:

* use RCrypto api to test bech32 with its plugin instead 
* expose this API via R2_USE_NEW_ABI, and prefix it with r_crypto_bech32_...

**Review by trufae — CHANGES_REQUESTED** [review]
. if you need more help better use the chat

**Comment by trufae** [comment]
please commit the missing file, i can take over the pr and do the needed fixes if you want, but the error in the build log complains about a file which i guess you didnt' commited

**Comment by W0nda** [comment]
Missing file coming.
Sorry for that

**Comment by trufae** [comment]
Another missing file
IMG_7349

**Comment by trufae** [comment]
just fixing it on a branch, but this code have quite many errors and warnings. ill fix them up in my pr.

```
p/crypto_bech32.c: In function ‘update’:
p/crypto_bech32.c:188:56: error: ‘enc’ undeclared (first use in this function)
  188 |                 bech32_encode (in_out, hrp, data, len, enc);
      |                                                        ^~~
p/crypto_bech32.c:188:56: note: each undeclared identifier is reported only once for each function it appears in
p/crypto_bech32.c:188:40: warning: passing argument 2 of ‘bech32_encode’ makes pointer from integer without a cast [-Wint-conversion]
  188 |                 bech32_encode (in_out, hrp, data, len, enc);
      |                                        ^~~
      |                                        |
      |                                        char
p/crypto_bech32.c:64:46: note: expected ‘const char *’ but argument is of type ‘char’
   64 | int bech32_encode (char *output, const char *hrp, const uint8_t *data, size_t data_len, bech32_encoding enc) {
      |                                  ~~~~~~~~~~~~^~~
p/crypto_bech32.c:188:45: warning: passing argument 3 of ‘bech32_encode’ makes pointer from integer without a cast [-Wint-conversion]
  188 |                 bech32_encode (in_out, hrp, data, len, enc);
      |                                             ^~~~
      |                                             |
      |                                             char
p/crypto_bech32.c:64:66: note: expected ‘const uint8_t *’ {aka ‘const unsigned char *’} but argument is of type ‘char’
   64 | int bech32_encode (char *output, const char *hrp, const uint8_t *data, size_t data_len, bech32_encoding enc) {
      |                                                   ~~~~~~~~~~~~~~~^~~~
p/crypto_bech32.c:188:51: warning: passing argument 4 of ‘bech32_encode’ makes integer from pointer without a cast [-Wint-conversion]
  188 |                 bech32_encode (in_out, hrp, data, len, enc);
      |                                                   ^~~
      |                                                   |
      |                                                   size_t * {aka long unsigned int *}
p/crypto_bech32.c:64:79: note: expected ‘size_t’ {aka ‘long unsigned int’} but argument is of type ‘size_t *’ {aka ‘long unsigned int *’}
   64 | int bech32_encode (char *output, const char *hrp, const uint8_t *data, size_t data_len, bech32_encoding enc) {
      |                                                                        ~~~~~~~^~~~~~~~
p/crypto_bech32.c:190:32: warning: passing argument 1 of ‘bech32_decode’ makes pointer from integer without a cast [-Wint-conversion]
  190 |                 bech32_decode (hrp, data, len, in_out);
      |                                ^~~
      |                                |
      |                                char
p/crypto_bech32.c:106:38: note: expected ‘char *’ but argument is of type ‘char’
  106 | bech32_encoding bech32_decode (char *hrp, uint8_t *data, size_t *data_len, const char *input) {
      |                                ~~~~~~^~~
p/crypto_bech32.c:190:37: warning: passing argument 2 of ‘bech32_decode’ makes pointer from integer without a cast [-Wint-conversion]
  190 |                 bech32_decode (hrp, data, len, in_out);
      |                                     ^~~~
      |                                     |
      |                                     char
p/crypto_bech32.c:106:52: note: expected ‘uint8_t *’ {aka ‘unsigned char *’} but argument is of type ‘char’
  106 | bech32_encoding bech32_decode (char *hrp, uint8_t *data, size_t *data_len, const char *input) {
      |                                           ~~~~~~~~~^~~~
p/crypto_bech32.c: In function ‘end’:
p/crypto_bech32.c:198:28: error: ‘hrp’ undeclared (first use in this function)
  198 |         return update (cj, hrp, data, data_len, in_out, enc);
      |                            ^~~
p/crypto_bech32.c:198:33: error: ‘data’ undeclared (first use in this function)
  198 |         return update (cj, hrp, data, data_len, in_out, enc);
      |                                 ^~~~
p/crypto_bech32.c:198:39: error: ‘data_len’ undeclared (first use in this function)
  198 |         return update (cj, hrp, data, data_len, in_out, enc);
      |                                       ^~~~~~~~
p/crypto_bech32.c:198:49: error: ‘in_out’ undeclared (first use in this function); did you mean ‘ino_t’?
  198 |         return update (cj, hrp, data, data_len, in_out, enc);
      |                                                 ^~~~~~
      |                                                 ino_t
p/crypto_bech32.c:198:57: error: ‘enc’ undeclared (first use in this function); did you mean ‘end’?
  198 |         return update (cj, hrp, data, data_len, in_out, enc);
      |                                                         ^~~
      |                                                         end
p/crypto_bech32.c:198:16: error: too many arguments to function ‘update’
  198 |         return update (cj, hrp, data, data_len, in_out, enc);
      |                ^~~~~~
p/crypto_bech32.c:184:13: note: declared here
  184 | static bool update(RCryptoJob *cj, const ut8 *buf, size_t *len) {
      |             ^~~~~~
p/crypto_bech32.c: At top level:
p/crypto_bech32.c:210:19: warning: initialization of ‘_Bool (*)(RCryptoJob *, const uint8_t *, int)’ {aka ‘_Bool (*)(struct r_crypto_job_t *, const unsigned char *, int)’} from incompatible pointer type ‘_Bool (*)(RCryptoJob *, const uint8_t *, size_t *)’ {aka ‘_Bool (*)(struct r_crypto_job_t *, const unsigned char *, long unsigned int *)’} [-Wincompatible-pointer-types]
  210 |         .update = update,
      |                   ^~~~~~
p/crypto_bech32.c:210:19: note: (near initialization for ‘r_crypto_plugin_bech32.update’)
p/crypto_bech32.c: In function ‘bech32_decode’:
p/crypto_bech32.c:169:1: warning: control reaches end of non-void function [-Wreturn-type]
  169 | }
      | ^
p/crypto_bech32.c: In function ‘end’:
p/crypto_bech32.c:199:1: warning: control reaches end of non-void function [-Wreturn-type]
  199 | }
      | ^

```

**Comment by trufae** [comment]
here you have: https://github.com/radareorg/radare2/pull/23528

cherry pick the last commit, as its the only one that is different from your branch, but even if it compiles this code wont work because it have lots of inconsistent issues and im not sure how to use the bech32 algorithm, what's hrp, adn so on. please update your pr with my commit and fix the code

**Comment by W0nda** [comment]
Hello,

The Bech32 algorithm is organized that way:
+-------+-------------+
| HRP | data part  |
+-------+-------------+

The problem with is that it requires 3 arguments:
1. bech32 string
2. bech32 string length
3. HRP size (can't be guessed)

while the update function in an RCryptoPlugin only takes
1. "buf"  (where bech32 string goes)
2. and "len" (where bech32 string length goes)
(and CryptoJob but that's not what's matter here).

Should I implement a whole new "update" function or is there another way to do it that is already in r2 ?

PS:
In crypto_bech32.c there is a BSD-3-Clause license, is it you that added that ?
If yes is it possible do replace it by a GPLv3 or not ?

**Comment by trufae** [comment]
Looks like you removed the message…Anyway answering from mailIirc the hrp must be provided by the user, this is passed as a key like its done in the aes plugin.About the license i just picked the first one i though you wanted but the code is yours so you can choose the one you like. The only problem is that if its gpl it will conflict with the lgpl one which means that must be guarded inside an ifdef and honor the without-gpl configure flag becauee gpl doesnt play well with static linkagesPing me on telegram, discord or matrix if you have questions—pancakeOn 26 Oct 2024, at 18:53, W0nda ***@***.***> wrote:﻿
Hello,
The Bech32 algorithm is organized that way:
+-------+-------------+
| HRP | data part  |
+-------+-------------+
The problem with is that it requires 3 arguments:

bech32 string
bech32 string length
HRP size (can't be guessed)

while the update function in an RCryptoPlugin only takes

"buf"  (where bech32 string goes)
and "len" (where bech32 string length goes)
(and CryptoJob but that's not what's matter here).

Should I implement a whole new "update" function or is there another way to do it that is already in r2 ?
PS:
In crypto_bech32.c there is a BSD-3-Clause license, is it you that added that ?
If yes is it possible do replace it by a GPLv3 or not ?

**Comment by W0nda** [comment]
I removed the message because I realised that it was wrong (HRP length is easily guessable).

Thank you for the info about how to pass the HRP.

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:23]: p/crypto_bech32.c:23: error: include file 'cstdio' not found
```suggestion
```

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:28]: No need to include any of these

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:33]: Use ut32 instead of uint32_t to follow the conventions

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:50]: Dont use assert. We dont want r2 apis to crash or use abort because they are used in libraries and programs cant control those exceptions

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:66]: Static

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:1]: Use one line license like all the other c files in r2, license must be in sync with the metadata in thebplugin struct below

**Review by trufae — CHANGES_REQUESTED** [review]
^

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:18]: p/crypto_bech32.c:18: error: ',' expected (got "enc")

^the bech32_encoding is not defined

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:4]: just include r_crypto.h and remove all those rlib, stdint, stdio.. because those includes are already included from rcrypto.h

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:24]: missingn return statement, i would recommend you to use a precondition and return 1 or 0 by default
```suggestion
		return 1;
	}
```

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:18]:
```suggestion
static uint32_t bech32_final_constant(bech32_encoding enc) {
	R_RETURN_VAL_IF_FAIL (enc ==  BECH32_ENCODING_BECH32 || enc == BECH32_ENCODING_BECH32M, 1);
```

File: libr/crypto/p/crypto_bech32.c
Comment (trufae) [inline libr/crypto/p/crypto_bech32.c:174]: check if hrp is null or not
```suggestion
	char *hrp = malloc (cj->key_len + 1); // HRP need to be null-terminated
	if (!hrp) {
		return false;
	}
```

File: libr/crypto/p/crypto_bech32.h
Comment (trufae) [inline libr/crypto/p/crypto_bech32.h:1]: header files dont need a license wall. only code needs that. because those are just interfaces and interfaces are not tied to licensing restrictions

File: libr/include/r_crypto/r_bech32.h
Comment (trufae) [inline libr/include/r_crypto/r_bech32.h:34]: you can probably move this enum into the crypto_bech32.c, dont expose new APIs, make everything accessilble thru the plugin

File: libr/main/rahash2.c
Comment (trufae) [inline libr/main/rahash2.c:406]: this will be fixed after the release. because it requires an abi breaking change, but for now i'm happy to merge it that way. thanks

**Review by trufae — CHANGES_REQUESTED** [review]
^

File: libr/crypto/p/crypto_bech32.c
Comment (W0nda) [reply libr/crypto/p/crypto_bech32.c:18]: bech32_encoding is defined in crypto_bech32.h.

There must be a problem I'll check that.

File: libr/include/r_crypto/r_bech32.h
Comment (W0nda) [reply libr/include/r_crypto/r_bech32.h:34]: OK I'll do that

**Comment by W0nda** [comment]
I forgot the precondition.

I change that and do another push

File: libr/crypto/p/crypto_bech32.c
Comment (W0nda) [reply libr/crypto/p/crypto_bech32.c:28]: If I follow all your suggested changes, this function will return 1 no matter what the "if" statements return,
is this what's expected ?

**Comment by trufae** [comment]
in case you didnt read the error:. `p/crypto_bech32.c:6: error: redefinition of enumerator 'BECH32_ENCODING_NONE'`

are you trying to compile this before pushing? do you want me to take over the PR instead?

**Comment by W0nda** [comment]
I was building with `sys/build.sh` that's why there was so many errors.

I builded with the right way this time and it should work.

I've another question: nothing left in `libr/include/r_crypto/r_bech32.h` because I removed the enum that was conflicting with `crypto_bech32.c`. Is there a way to remove this file or do I have to keep it (I can't build without it for the moment) ?

**Comment by trufae** [comment]
```
p/crypto_bech32.c: In function ‘update’:
p/crypto_bech32.c:177:6: error: unused variable ‘hrp_size’ [-Werror=unused-variable]
  177 |  int hrp_size = hrplength (buf, len);
      |      ^~~~~~~~
p/crypto_bech32.c: In function ‘bech32_final_constant’:
p/crypto_bech32.c:29:1: error: control reaches end of non-void function [-Werror=return-type]
   29 | }
      | ^
rm -f ios-arm-64.sdb
```

File: libr/include/r_crypto/r_bech32.h
Comment (trufae) [inline libr/include/r_crypto/r_bech32.h:1]: i guess we can remove this file completely, because it doesnt provide any define or enum

File: libr/include/r_crypto.h
Comment (trufae) [inline libr/include/r_crypto.h:165]:
```suggestion
#define R_CRYPTO_BECH32 1ULL << 15
```

unnecessary space

**Comment by trufae** [comment]
you can build with meson or ./configure ; make, no need to do full clean builds all the time. sys/build.sh is actually executed from sys/install.sh which is the recommended clean build script. so it should work well too. ill merge after those minor fixes. thank you and congrats for your first contribution!

**Review by trufae — CHANGES_REQUESTED** [review]
^

File: libr/crypto/p/crypto_bech32.c
Comment (W0nda) [reply libr/crypto/p/crypto_bech32.c:208]: yes

**Comment by W0nda** [comment]
I can't build without r_bech32.h (that's why I left it last commit).

**Comment by trufae** [comment]
if you cant build withhout the r_bech32.h is because its included or referenced from a .d somewhere.

**Comment by W0nda** [comment]
Apparently I can build without r_bech32.h ...

**Comment by W0nda** [comment]
I tried to resolve the conflicts i hope it worked.

**Comment by trufae** [comment]
Loooks like building now! Good job!

**Pull request merged**
